File: src/main/java/com/lyaiya/softcode/mixin/SoftcodeLateMixinLoader.java
package com.lyaiya.softcode.mixin;

import com.lyaiya.softcode.Tags;
import com.lyaiya.softcode.constant.ModIdConstant;
import net.minecraftforge.fml.common.Loader;
import zone.rong.mixinbooter.ILateMixinLoader;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SoftcodeLateMixinLoader implements ILateMixinLoader {
    private final Map<String, String> mixinConfigs = new HashMap<>();

    public SoftcodeLateMixinLoader() {
        put(mixinConfigs, ModIdConstant.IN_WORLD_CRAFTING);
        put(mixinConfigs, ModIdConstant.EX_NIHILO_CREATIO);
        put(mixinConfigs, ModIdConstant.THAUMCRAFT);
        put(mixinConfigs, ModIdConstant.THAUMIC_JEI);
        put(mixinConfigs, ModIdConstant.IMMERSIVE_ENGINEERING);
        put(mixinConfigs, ModIdConstant.BLOCK_DROPS);
        put(mixinConfigs, ModIdConstant.IC2);
        put(mixinConfigs, ModIdConstant.CONTROLLING);
        put(mixinConfigs, ModIdConstant.LIB_VULPES);
        put(mixinConfigs, ModIdConstant.ADVANCED_ROCKETRY);
        put(mixinConfigs, ModIdConstant.SERENES_SEASONS);
        put(mixinConfigs, ModIdConstant.JUST_ENOUGH_RESOURCES);
        put(mixinConfigs, ModIdConstant.INDUSTRIAL_FOREGOING);
    }

    private void put(Map<String, String> map, String modid) {
        map.put(String.format("mixins.%s.%s.json", Tags.MOD_ID, modid), modid);
    }

    @Override
    public List<String> getMixinConfigs() {
        return new ArrayList<>(mixinConfigs.keySet());
    }

    @Override
    public boolean shouldMixinConfigQueue(String mixinConfig) {
        return mixinIfModLoaded(mixinConfig);
    }

    public boolean mixinIfModLoaded(String mixinConfig) {
        if (mixinConfigs.containsKey(mixinConfig)) {
            final String modid = mixinConfigs.get(mixinConfig);
            if (modid != null) {
                return Loader.isModLoaded(modid);
            }
        }
        return false;
    }

}

File: src/main/java/com/lyaiya/softcode/mixin/SoftcodeLateMixinLoader.kt
package com.lyaiya.softcode.mixin

import com.lyaiya.softcode.Softcode
import com.lyaiya.softcode.constant.ModIdConstant
import net.minecraftforge.fml.common.Loader
import zone.rong.mixinbooter.ILateMixinLoader

class SoftcodeLateMixinLoader : ILateMixinLoader {
    private val mixinJsons = buildMap {
        putMixinJson(ModIdConstant.IN_WORLD_CRAFTING)
        putMixinJson(ModIdConstant.EX_NIHILO_CREATIO)
        putMixinJson(ModIdConstant.THAUMCRAFT)
        putMixinJson(ModIdConstant.THAUMIC_JEI)
        putMixinJson(ModIdConstant.IMMERSIVE_ENGINEERING)
        putMixinJson(ModIdConstant.BLOCK_DROPS)
        putMixinJson(ModIdConstant.IC2)
        putMixinJson(ModIdConstant.CONTROLLING)
        putMixinJson(ModIdConstant.LIB_VULPES)
        putMixinJson(ModIdConstant.ADVANCED_ROCKETRY)
        putMixinJson(ModIdConstant.SERENES_SEASONS)
        putMixinJson(ModIdConstant.JUST_ENOUGH_RESOURCES)
        putMixinJson(ModIdConstant.INDUSTRIAL_FOREGOING)
    }

    override fun getMixinConfigs(): List<String> {
        return mixinJsons.values.toList()
    }

    override fun shouldMixinConfigQueue(mixinConfig: String): Boolean {
        return mixinIfModLoaded(mixinConfig)
    }

    private fun mixinIfModLoaded(mixinConfig: String): Boolean {
        for ((modid, mixinJson) in mixinJsons) {
            if (mixinConfig == mixinJson) {
                return Loader.isModLoaded(modid)
            }
        }
        return false
    }

    private fun MutableMap<String, String>.putMixinJson(modid: String) {
        this[modid] = "mixins.${Softcode.MOD_ID}.${modid}.json"
    }
}